File: decorator/java/GeometricForm.java
interface Shape {

	void draw();
}

final class Circle implements Shape {

	@Override
	public void draw() {
		System.out.println("Draw circle");
	}
}

final class Ractangle implements Shape {

	@Override
	public void draw() {
		System.out.println("Draw rectangle");
	}
}

abstract class ShapeDecorator implements Shape {

	private Shape shape;

	protected ShapeDecorator(Shape shape) {
		this.shape = shape;
	}

	@Override
	public void draw() {
		shape.draw();
	}
}

final class CircleDecorator extends ShapeDecorator {

	CircleDecorator(Shape shape) {
		super(shape);
	}

	@Override
	public void draw() {
		super.draw();
		this.addRedBorder();
	}

	private void addRedBorder() {
		System.out.println("Add border red to circle");
	}
}

final class RectangleDecorator extends ShapeDecorator {

	RectangleDecorator(Shape shape) {
		super(shape);
	}

	@Override
	public void draw() {
		super.draw();
		this.addRedBorder();
	}

	private void addRedBorder() {
		System.out.println("Add border red to rectangle");
	}
}

final class GeometricForm {

	public static void main(String[] args) {
		Shape shape = new Circle();
		ShapeDecorator shapeDecorator = new CircleDecorator(shape);

		shapeDecorator.draw();

		shape = new Ractangle();
		shapeDecorator = new CircleDecorator(shape);

		shapeDecorator.draw();
	}
}

File: decorator/kotlin/geometric_form.kt
interface Shape {

	fun draw(): Unit
}

class Circle : Shape {

	override fun draw(): Unit {
		println("Draw circle");
	}
}

class Rectangle : Shape {

	override fun draw(): Unit {
		println("Draw rectangle");
	}
}

abstract class AbstractDecorator(private val shape: Shape) : Shape {

	override fun draw(): Unit {
		shape.draw()
	}
}

class CircleDecorator : AbstractDecorator {

	constructor(shape: Shape) : super(shape)

	override fun draw(): Unit {
		super.draw()
		this.addRedBorder()
	}

	private fun addRedBorder(): Unit {
		println("Add red border to circle");
	}
}

class RectangleDecorator : AbstractDecorator {

	constructor(shape: Shape) : super(shape)

	override fun draw(): Unit {
		super.draw()
		this.addRedBorder()
	}

	private fun addRedBorder(): Unit {
		println("Add red border to circle");
	}
}

fun main(): Unit {
	val circle: Circle = Circle()
	val rectangle: Rectangle = Rectangle()

	// Decorates the circle
	var decorator: AbstractDecorator = CircleDecorator(circle)

	decorator.draw()

	// Decorates the rectangle
	decorator = RectangleDecorator(rectangle)

	decorator.draw()
}